// react-static/lib/static/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "extractTemplates", {
  enumerable: true,
  get: function get() {
    return _extractTemplates.default;
  }
});
Object.defineProperty(exports, "preparePlugins", {
  enumerable: true,
  get: function get() {
    return _preparePlugins.default;
  }
});
Object.defineProperty(exports, "prepareRoutes", {
  enumerable: true,
  get: function get() {
    return _prepareRoutes.default;
  }
});
Object.defineProperty(exports, "fetchSiteData", {
  enumerable: true,
  get: function get() {
    return _fetchSiteData.default;
  }
});
Object.defineProperty(exports, "exportSharedRouteData", {
  enumerable: true,
  get: function get() {
    return _exportSharedRouteData.default;
  }
});
Object.defineProperty(exports, "buildHTML", {
  enumerable: true,
  get: function get() {
    return _buildHTML.default;
  }
});
Object.defineProperty(exports, "generateTemplates", {
  enumerable: true,
  get: function get() {
    return _generateTemplates.default;
  }
});
Object.defineProperty(exports, "generateBrowserPlugins", {
  enumerable: true,
  get: function get() {
    return _generateBrowserPlugins.default;
  }
});
Object.defineProperty(exports, "fetchRoutes", {
  enumerable: true,
  get: function get() {
    return _fetchRoutes.default;
  }
});
Object.defineProperty(exports, "exportRoute", {
  enumerable: true,
  get: function get() {
    return _exportRoute.default;
  }
});
Object.defineProperty(exports, "exporter", {
  enumerable: true,
  get: function get() {
    return _exporter.default;
  }
});
Object.defineProperty(exports, "buildXML", {
  enumerable: true,
  get: function get() {
    return _buildXML.default;
  }
});
Object.defineProperty(exports, "exportRoutes", {
  enumerable: true,
  get: function get() {
    return _exportRoutes.default;
  }
});
Object.defineProperty(exports, "getConfig", {
  enumerable: true,
  get: function get() {
    return _getConfig.default;
  }
});
Object.defineProperty(exports, "reloadRoutes", {
  enumerable: true,
  get: function get() {
    return _webpack.reloadRoutes;
  }
});
Object.defineProperty(exports, "startDevServer", {
  enumerable: true,
  get: function get() {
    return _webpack.startDevServer;
  }
});
Object.defineProperty(exports, "buildProductionBundles", {
  enumerable: true,
  get: function get() {
    return _webpack.buildProductionBundles;
  }
});

var _extractTemplates = _interopRequireDefault(require("./extractTemplates"));

var _preparePlugins = _interopRequireDefault(require("./preparePlugins"));

var _prepareRoutes = _interopRequireDefault(require("./prepareRoutes"));

var _fetchSiteData = _interopRequireDefault(require("./fetchSiteData"));

var _exportSharedRouteData = _interopRequireDefault(require("./exportSharedRouteData"));

var _buildHTML = _interopRequireDefault(require("./buildHTML"));

var _generateTemplates = _interopRequireDefault(require("./generateTemplates"));

var _generateBrowserPlugins = _interopRequireDefault(require("./generateBrowserPlugins"));

var _fetchRoutes = _interopRequireDefault(require("./fetchRoutes"));

var _exportRoute = _interopRequireDefault(require("./exportRoute"));

var _exporter = _interopRequireDefault(require("./exporter"));

var _buildXML = _interopRequireDefault(require("./buildXML"));

var _exportRoutes = _interopRequireDefault(require("./exportRoutes"));

var _getConfig = _interopRequireDefault(require("./getConfig"));

var _webpack = require("./webpack");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
